src/server_name.rs

```diff
@@ -204,6 +204,14 @@ impl<'a> DnsName<'a> {
             Err(_) => Err(s),
         }
     }
+
+    /// Produces a borrowed [`DnsName`] from a borrowed [`str`].
+    pub const fn try_from_str(s: &str) -> Result<DnsName<'_>, InvalidDnsNameError> {
+        match validate(s.as_bytes()) {
+            Ok(_) => Ok(DnsName(DnsNameInner::Borrowed(s))),
+            Err(_) => Err(InvalidDnsNameError),
```

Comment:
Suggest yielding the error returned by `validate()` rather than reinstantiating the same value.

Can we replace the implementation in `TryFrom<&str>` with a call to this function to avoid duplication?